### PR TITLE
=ht2 refactor multiplexer side of demux into its own class

### PR DESCRIPTION
**Pull request by jrudolph**
This is groundwork for upcoming changes. The multiplexer now has a defined interface. This will helps with code organization but might also allow to try out alternative implementations.

### CODE REVIEW

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/BufferedOutletSupport.scala [hunk 1]
```diff
@@ -8,85 +8,97 @@ import akka.stream.Outlet
...
-    }
+  override def onPull(): Unit =
+    if (!buffer.isEmpty) doPush(buffer.pop())
 
-    override def onPull(): Unit =
-      if (!buffer.isEmpty) doPush(buffer.pop())
+  outlet.setHandler(this)
 
-    outlet.setHandler(this)
+  final def push(element: T): Unit = pushWithTrigger(element, () ⇒ ())
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/BufferedOutletSupport.scala]: cache the `() => ()`?

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * The internal interface Http2ServerDemux uses to drive the multiplexer.
+ */
+private[http2] trait Http2Multiplexer {
+  def pushControlFrame(frame: FrameEvent): Unit
+  def registerSubStream(sub: Http2SubStream): Unit
+  def cancelSubStream(streamId: Int): Unit
+  def updateWindow(streamId: Int, increment: Int): Unit
+  def updateFrameSize(newFrameSize: Int): Unit
+  def updateDefaultWindow(newDefaultWindow: Int): Unit
+}
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala]: nice to give it clean API like this 👍

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      }
+      def updateWindow(streamId: Int, increment: Int): Unit = {
+        if (streamId == 0) {
+          totalOutboundWindowLeft += increment
+          debug(s"Updating outgoing connection window by $increment to $totalOutboundWindowLeft")
+        } else {
+          updateWindowFor(streamId, increment)
+          debug(s"Updating window for $streamId by $increment to ${windowLeftFor(streamId)} bufferedElements: ${buffer.size()}")
+        }
+        tryFlush()
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala]: much cleaner with this refactor, nice 👍

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala [hunk 4]
```diff
@@ -88,17 +88,35 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
 
       override def doPush(elem: ElementAndTrigger): Unit =
         elem.element match {
-          case d @ DataFrame(streamId, _, pl) ⇒
-            if (pl.size <= totalOutboundWindowLeft && pl.size <= windowLeftFor(streamId)) {
+          case d @ DataFrame(streamId, _, payload) ⇒
+            val sendSize = totalOutboundWindowLeft min windowLeftFor(streamId) min payload.size
+
+            def send(elem: ElementAndTrigger): Unit = {
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala]: This seems to duplicate what I had implemented in FrameRendering?

Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala]: Ah I see, this is handling window size, whereas the renderer was handling frame size limits. 
Good as is!

**Comment by jrudolph** [comment]
PLS BUILD

**Comment by ktoso** [comment]
LGTM

**Pull request merged**